crates/bitwarden-core/src/key_management/crypto.rs

```diff
@@ -574,45 +583,97 @@ pub(super) fn verify_asymmetric_keys(
+    /// User key
+    user_key: String,
+
+    /// Wrapped private key
+    private_key: EncString,
+    /// Public key
+    public_key: String,
     /// The user's public key, signed by the signing key
     signed_public_key: SignedPublicKey,
+
+    /// Signing key, encrypted with the user's symmetric key
+    signing_key: EncString,
+    /// Base64 encoded verifying key
+    verifying_key: String,
+
+    /// The user's signed security state
+    security_state: SignedSecurityState,
+    /// The security state's version
+    security_version: u64,
 }
 
-/// Makes a new set of signing keys for a user, which should only be done during
-/// once. This also signs the public key with the signing key
-/// and returns the signed public key.
-pub fn make_user_signing_keys_for_enrollment(
+/// Initializes the user's cryptographic state for v2 users.
+/// If the client already contains a v1 user, then this user's private-key will be
+/// re-used.
+pub(crate) fn make_keys_for_user_crypto_v2(
     client: &Client,
-) -> Result<MakeUserSigningKeysResponse, CryptoError> {
+) -> Result<EnrollUserCryptoV2Response, CryptoError> {
     let key_store = client.internal.get_key_store();
     let mut ctx = key_store.context();
 
-    // Make new keypair and sign the public key with it
-    let signature_keypair = SigningKey::make(SignatureAlgorithm::Ed25519);
-    let temporary_signature_keypair_id = SigningKeyId::Local("temporary_key_for_rotation");
+    let temporary_user_key_id = SymmetricKeyId::Local("temporary_user_key");
+    let temporary_signing_key_id = SigningKeyId::Local("temporary_signing_key");
+    let temporary_private_key_id = AsymmetricKeyId::Local("temporary_private_key");
+
+    // If the user already has a private key, use it. Otherwise, create a temporary one.
+    let private_key_id = if ctx.has_asymmetric_key(AsymmetricKeyId::UserPrivateKey) {
+        AsymmetricKeyId::UserPrivateKey
+    } else {
+        ctx.make_asymmetric_key(temporary_private_key_id)?
+    };
+    #[allow(deprecated)]
+    let private_key = ctx.dangerous_get_asymmetric_key(private_key_id)?.clone();
+
+    // New user key
+    let user_key = SymmetricCryptoKey::make_xchacha20_poly1305_key();
     #[allow(deprecated)]
-    ctx.set_signing_key(temporary_signature_keypair_id, signature_keypair.clone())?;
-    let signed_public_key = ctx.make_signed_public_key(
-        AsymmetricKeyId::UserPrivateKey,
-        temporary_signature_keypair_id,
+    ctx.set_symmetric_key(temporary_user_key_id, user_key.clone())?;
+
+    // New signing key
+    let signing_key = SigningKey::make(SignatureAlgorithm::Ed25519);
+    #[allow(deprecated)]
+    ctx.set_signing_key(temporary_signing_key_id, signing_key.clone())?;
+
+    // Sign existing public key
+    let signed_public_key = ctx.make_signed_public_key(private_key_id, temporary_signing_key_id)?;
+    #[allow(deprecated)]
+    let public_key = ctx
+        .dangerous_get_asymmetric_key(private_key_id)?
+        .to_public_key();
```

Comment:
We already get the private key in line 633, right?
```suggestion
    let public_key = private_key.to_public_key();
```